ocaml/xapi/xapi_session.ml

```diff
@@ -27,6 +27,232 @@ open Client
+%s
+<number>%i</number>
+<date>%s</date>
+</known>|}
+      (string_of_client x.client)
+      x.num_failed_attempts
+      (Date.to_string x.last_failed_attempt)
+
+  type stats = {
+      total_num_failed_attempts: int
+    ; top_3_worst_clients: client_failed_attempts list
+          (* not necessarily 3, but <=3 *)
+    ; unknown_client_failed_attempts: int
+  }
+
+  let string_of_stats
+      {
+        total_num_failed_attempts
+      ; top_3_worst_clients
+      ; unknown_client_failed_attempts
+      } =
+    let unknown =
+      if unknown_client_failed_attempts = 0 then
+        ""
+      else
+        Printf.sprintf {|
+<unknown>%i</unknown>|} unknown_client_failed_attempts
+    in
+    let known_with_total =
+      if List.length top_3_worst_clients = 0 then
+        ""
+      else
+        Printf.sprintf {|
+<total>%i</total>%s|} total_num_failed_attempts
+          (top_3_worst_clients
+          |> List.map string_of_client_failed_attempts
+          |> String.concat ""
+          )
+    in
+    Printf.sprintf {|<body>%s%s
+</body>|} known_with_total unknown
+
+  module Stats : sig
+    val get : unit -> stats option
+
+    (* returns the number of failures from this client since last call to [ get ] *)
+    val record_client : client -> now:Date.iso8601 -> int
+
+    (* returns number of failures from unknown clients since last call to [ get ] *)
+    val record_unknown : unit -> int
+  end = struct
+    let m = Mutex.create ()
+
+    let unknown_ctr = ref 0
+
+    let record_unknown () =
+      Mutex.execute m (fun () ->
+          let ctr = !unknown_ctr + 1 in
+          unknown_ctr := ctr ;
+          ctr)
+
+    type key = client
+
+    type value = {num_failed_attempts: int; last_failed_attempt: Date.iso8601}
+
+    type table_t = (key, value) Hashtbl.t
+
+    let table = Hashtbl.create 10
+
+    let record_client k ~now =
+      Mutex.execute m (fun () ->
+          match Hashtbl.find_opt table k with
+          | None ->
+              Hashtbl.add table k
+                {num_failed_attempts= 1; last_failed_attempt= now} ;
+              1
+          | Some ({num_failed_attempts} : value) ->
+              let num_failed_attempts = num_failed_attempts + 1 in
+              Hashtbl.replace table k
+                {num_failed_attempts; last_failed_attempt= now} ;
+              num_failed_attempts)
+
+    let get () =
+      let reset () =
+        Hashtbl.reset table ;
+        unknown_ctr := 0
+      in
+      Mutex.execute m (fun () ->
+          let unknown_client_failed_attempts = !unknown_ctr in
+          if Hashtbl.length table = 0 && unknown_client_failed_attempts = 0 then
+            None
+          else
+            let num_known_client_failed_attempts, top_3_worst_clients =
+              Hashtbl.fold
+                (fun client {num_failed_attempts; last_failed_attempt}
+                     (ctr, worst_so_far) ->
+                  ( ctr + num_failed_attempts
+                  , up_to_3 worst_so_far
```

Comment:
This looks lineal to me, and only called when the values need to be collected for the alerts, which seems like a good compromise to me.